### PR TITLE
Code review

### PR DESCRIPTION
**Pull request by varunrajk**
Hi MDP Developers, 

Here is my first pull request. 

It includes only an OnlineNode, OnlineFlow and CircularFlow. I added corresponding test functions for each and also two online nodes (Moving average and Moving time diff) as basic online nodes. 

Looking forward to your edits and comments!

-Varun

### CODE REVIEW

File: mdp/linear_flows_online.py [hunk 1]
```diff
@@ -0,0 +1,607 @@
...
+                _cache['node#%d' % i] = node.cache
+        return _cache
+
+    def __setitem__(self, key, value):
+        if isinstance(key, slice):
+            [self._check_value_type_is_compatible(item) for item in value]
+        else:
+            self._check_value_type_is_compatible(value)
+
+        # make a copy of list
```
Comment (nwilbert) [inline mdp/linear_flows_online.py]: I'd leave out comments like these that just describe what the code is doing (without providing real value).

File: mdp/linear_flows_online.py [hunk 2]
```diff
@@ -0,0 +1,607 @@
...
+
+class CircularOnlineFlowException(mdp.MDPException):
+    """Base class for exceptions in Flow subclasses."""
+    pass
+
+
+class CircularOnlineFlow(OnlineFlow):
+    """A 'CircularOnlineFlow' is a cyclic sequence of online/non-trainable nodes that are trained and executed
+    together to form a more complex algorithm.  The input and output nodes of the flow can be changed
+    at any time using the functions "set_input_node" and "set_output_node".
```
Comment (nwilbert) [inline mdp/linear_flows_online.py]: This seems like a finer detail, that could be placed further down.

File: mdp/test/test_onlinenode.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    flow = BogusOnlineNode() + BogusOnlineNode()
+    assert(isinstance(flow, mdp.OnlineFlow))
+    inp = mdp.numx.zeros([2,5])
+    flow.train(inp)
+    out = flow.execute(inp)
+    assert_array_equal(mdp.numx.ones(inp.shape)*2, out)
+    for node in flow:
+        assert(node.is_training())
+
+    flow = BogusOnlineNode() + BogusNodeTrainable()
```
Comment (nwilbert) [inline mdp/test/test_onlinenode.py]: Shouldn't this be a separate test?

File: mdp/test/test_online_flows_hinet.py [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+    flow.append(newnode)
+    assert_equal(len(flow), length+1)
+    length = len(flow)
+    try:
+        newnode = BogusNode(input_dim=11)
+        flow.append(newnode)
+        raise Exception('flow.append appended inconsistent node')
+    except ValueError:
+        assert_equal(len(flow), length)
+    # extend
```
Comment (nwilbert) [inline mdp/test/test_online_flows_hinet.py]: I'd split this into separate tests.

File: mdp/linear_flows_online.py [hunk 5]
```diff
@@ -0,0 +1,607 @@
...
+    first node and is successively processed by the subsequent nodes along
+    the sequence.
+
+    Using an online flow as opposed to handling manually a set of nodes has a
+    clear advantage: The general online flow implementation automatizes the
+    training (including supervised training and multiple training phases),
+    execution, and inverse execution (if defined) of the whole sequence.
+
+    OnlineFlow sequence can contain trained or non-trainable Nodes and optionally
+    a terminal trainable Node.
```
Comment (nwilbert) [inline mdp/linear_flows_online.py]: I think we generally have to be a little careful with the term "training" and "trainable". On the one hand there are "online" nodes that "learn" in their executable state (i.e., with node.execute). On the other hand there are "trainable" nodes, that have standard training phases.
Here's my current understanding (please correct me if this is off): What OnlineFlow really does is to enable specific additional data (supervision signals) to be passed to each node in the flow during the training of the last node in the flow (that is allowed to be trainable).

File: mdp/linear_flows_online.py [hunk 6]
```diff
@@ -0,0 +1,607 @@
...
+       number of nodes or a single numpy array. OnlineFlow requires only one
+       input dataiterable as each node is trained simultaneously.
+
+    c) Additional train args (supervised labels etc) are passed to each node through the
+       node specific dataiterable. OnlineFlow requires the dataiterable to return a list
+       that contains tuples of args for each node: [x, (node0 args), (node1 args), ...]. See
+       train docstring.
+
+    d) OnlineFlow also has the cache attribute that retrieves any caches variables
+       stored in the OnlineNodes.
```
Comment (nwilbert) [inline mdp/linear_flows_online.py]: What does this cache exactly do? Is it coupled to the concept of online learning, or is here because it was useful for some specific case?

File: mdp/linear_flows_online.py (hunk 1 above)
Comment (varunrajk) [inline mdp/linear_flows_online.py]: Sure.

**Comment by varunrajk** [comment]
I seemed to have messed up the branch and pull request while working on the edits. I should hopefully be able to restore it back.

File: mdp/test/test_online_flows_hinet.py (hunk 4 above)
Comment (varunrajk) [inline mdp/test/test_online_flows_hinet.py]: I simply added this test from 'test_flows.py' with modifications to include online flow. In any case, here each list method test uses common values. So, splitting them would require rewriting all the tests.

File: mdp/linear_flows_online.py (hunk 5 above)
Comment (varunrajk) [inline mdp/linear_flows_online.py]: From reverse engineering mdp :) Here is my understanding between a few terms (correct me if I am wrong):
trainable - This defines whether a node is trainable or not. That is, is_trainable() returns True.
training - This is actually a state of a trainable node. If training completes is_training() returns False. 

A "trainable" node implies that is_trainable() = True. It does not indicate the current state of the node, that is, the training may have been finished or not. is_training() = True or False.
A "non-trainable" node implies that is_trainable() returns False and so does is_training().
A "trained" node implies that is_trainable() returns True, but is_training() returns False. 

So, an online flow could be a sequence of 
(a) OnlineNodes, or
(b) a mix of OnlineNodes and "trained" or "non-trainable" nodes, or
(c) a mix of OnlineNodes/trained/non-trainable nodes and a terminal "trainable" node. 

An OnlineFlow trains all the nodes simultaneously for the given input sample x. 
So, node2 is training while node1 is still training (node1.is_training() = True)

Whereas a standard Flow trains nodes sequentially, that is, the second node is only trained after the first node's training has finished  (node1.is_training() = False)

I have tried to make sure that online nodes only "learn" when their train method is called. execute method does not change the state of the algorithm only train method does.

Comment (varunrajk) [inline mdp/linear_flows_online.py]: Please let me know if this makes sense. Then I can improve the doc string of the Online flow to disambiguate some of the terms used.

File: mdp/linear_flows_online.py (hunk 6 above)
Comment (varunrajk) [inline mdp/linear_flows_online.py]: Yes, cache can store some intermediate variables that are not an output via an execute call. These can be useful for debugging/plotting etc. User can choose the variables to monitor during training.

File: mdp/nodes/stats_nodes_online.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+     Also supports exponentially weighted moving average when
+     the parameter avg_n is set.
+
+    This is an online learnable node (OnlineNode)
+
+    **Internal variables of interest (stored in cache)**
+
+      ``self.avg``
+          The current average of the input data
+    """
```
Comment (pberkes) [inline mdp/nodes/stats_nodes_online.py]: You need to document what the execution does.

File: mdp/nodes/stats_nodes_online.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
+import mdp
+
+
+class MovingAvgNode(mdp.PreserveDimOnlineNode):
```
Comment (pberkes) [inline mdp/nodes/stats_nodes_online.py]: If I understand correctly, this is not really a "moving average" in the usual sense (e.g. https://en.wikipedia.org/wiki/Moving_average). It's more of an online average calculation, so I would suggest to call it "OnlineMeanNode", "OnlineAvgNode", or similar

File: mdp/nodes/stats_nodes_online.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+            typ = 'dtype=None'
+        else:
+            typ = "dtype='%s'" % self.dtype.name
+        numx_rng = "numx_rng=%s" % str(self.numx_rng)
+        avg_n = "avg_n=%s" % self.avg_n
+        args = ', '.join((inp, out, typ, numx_rng, avg_n))
+        return name + '(' + args + ')'
+
+
+class MovingTimeDiffNode(mdp.PreserveDimOnlineNode):
```
Comment (pberkes) [inline mdp/nodes/stats_nodes_online.py]: This is also not "moving" in the sense of "computed in a moving window". Maybe again "OnlineTimeDiffNode"?

**Comment by pberkes** [comment]
There's a conflict that stops Travis CI from running.

**Comment by pberkes** [comment]
Thanks fo the PR @varunrajk ! I added my review

File: mdp/linear_flows_online.py [hunk 10]
```diff
@@ -0,0 +1,658 @@
...
+    """
+
+    def __init__(self, flow, crash_recovery=False, verbose=False):
+        super(OnlineFlow, self).__init__(flow, crash_recovery, verbose)
+        # check if the list of nodes is compatible.
+        self._check_compatibilitiy(flow)
+        # collect cache from individual nodes.
+        self._cache = self._get_cache_from_flow(flow)
+        # collect train_args for each node
+        self._train_arg_keys_list, self._train_args_needed_list = self._get_required_train_args_from_flow(flow)
```
Comment (varunrajk) [inline mdp/linear_flows_online.py]: Yes it is needed and is checked for OnlineFlows too. OnlineFlow does not overwrite  "_check_nodes_consistency" method from its parent class.

File: mdp/linear_flows_online.py [hunk 11]
```diff
@@ -0,0 +1,658 @@
...
+    def _check_value_type_is_compatible(self, value):
+        # onlinenodes, trained and non-trainable nodes are compatible
+        if not isinstance(value, mdp.Node):
+            raise TypeError("flow item must be a Node instance and not %s" % type(value))
+        elif isinstance(value, mdp.OnlineNode):
+            pass
+        else:
+            # classic mdp Node
+            if value.is_training():
+                raise TypeError("flow item must either be an OnlineNode instance, a trained or a non-trainable Node.")
```
Comment (varunrajk) [inline mdp/linear_flows_online.py]: I think there seems to be a confusion with the compatible node sequences for an OnlineFlow. I improved the docstrings and also renamed a method to make this clear. To answer your question let me first discuss the compatible node sequences:

OnlineFlow node sequence can contain
 (a) only OnlineNodes
        (Eg. [MovingAvgNode(), IncSFANode()],
 or
(b) a mix of OnlineNodes and trained/non-trainable Nodes
        (eg. [a fully trained PCANode, IncSFANode()] or [QuadraticExpansionNode(), IncSFANode()],
or
 (c) a mix of OnlineNodes/trained/non-trainable Nodes and a terminal trainable Node (but not an OnlineNode) whose training hasn't finished
        (eg. [IncSFANode(), QuadraticExpansionNode(), a partially or untrained SFANode]).

(a) is fairly obvious. (b) allows inclusion of non-trainable or pre-trained nodes. For example, a pretrained three layered hierarchical SFA node and a top incremental SFA node. (c) allows an uncommon combination of online preprocessing for training a batch node. 

The confusion could also stem from the method names I chose before:
"_check_value_type_is_compatible" (an interim compatibility check)
and 
"_check_compatibilitiy" (this is the actual one)

 I changed these to 
"_check_value_type_is_online_or_nontrainable_node" (renamed)
and 
"_check_compatibilitiy" (same as before).

I hope this clarifies your doubt. Thanks.

File: mdp/linear_flows_online.py [hunk 12]
```diff
@@ -0,0 +1,658 @@
...
+
+    def _check_compatibilitiy(self, flow):
```
Comment (varunrajk) [inline mdp/linear_flows_online.py]: Thanks.

File: mdp/test/__init__.py [hunk 13]
```diff
@@ -27,7 +27,8 @@ def test(filename=None, seed=None, options='', mod_loc=None):
     # add --assert=reiterp option to work around permissions problem
     # with __pycache__ directory when MDP is installed on a normal
     # user non-writable directory
-    options = "--assert=reinterp "+options
+    # options = "--assert=reinterp "+options
+    options = "-s "+options
```
Comment (otizonaizit) [inline mdp/test/__init__.py]: I think here the problem is that you did not merge master on your branch recently? The "assert=reinterp" option to py.test has been removed with the last commit in master: d44efa5 . As that was a revert commit, it creates a conflict. Damn, next time I should simply revert manually and create a new commit...

File: mdp/nodes/stats_nodes_online.py (hunk 8 above)
Comment (varunrajk) [inline mdp/nodes/stats_nodes_online.py]: There are two parts to this comment:

First on the definition of the moving average:
MovingAvgNode supports two averages: (a) A cumulative moving average according to the wiki page you have pointed out, which is also an online average calculation. (b) An exponentially weighted moving average. All moving averages more or less follow an update equation similar to:
avg_t = alpha * x_t + (1-alpha) * avg_t-1. For cumulative moving avg: alpha = 1/t and for exponential it is a constant. So I believe in this regard the usage of moving average as its name is justified. 

 However, after taking a second look, I figured out that this may not be a right name for the node after all. Since the node does not return an average but returns the centered data. So, I aptly changed the name of the node to OnlineCenteringNode and improved the docstring. The node's main purpose is to center the data via an execute call by updating an average via the train call. 

Thanks for pointing this out anyway :)

File: mdp/nodes/stats_nodes_online.py (hunk 9 above)
Comment (varunrajk) [inline mdp/nodes/stats_nodes_online.py]: Yes, I completely agree with you on this. Changed it to OnlineTimeDiffNode and improved the doc string.

**Comment by pberkes** [comment]
Hi @varunrajk . I'm happy with the revision. The only thing I'm concerned about it is commit https://github.com/mdp-toolkit/mdp-toolkit/pull/20/commits/a4910b987a4956c34c965ad6a44caab9811cc949 : Was it a bug as the commit title suggests? Then please add a test reproducing the bug.

**Comment by varunrajk** [comment]
Hi @pberkes, Thanks for the approval. I added the necessary test function to check for the bug.

File: mdp/nodes/stats_nodes_online.py [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+    def __init__(self, avg_n=None, input_dim=None, output_dim=None, dtype=None, numx_rng=None):
+        """
+        avg_n - If set to None (default), the node updates a simple moving average. If set to a positive integer,
+        the node updates an exponentially weighted moving average.
+        """
+        super(OnlineCenteringNode, self).__init__(input_dim=input_dim, output_dim=output_dim, dtype=dtype,
+                                                  numx_rng=numx_rng)
+        self.avg_n = avg_n
+        self.avg = None
+        self._cache = {'avg': None}
```
Comment (nwilbert) [inline mdp/nodes/stats_nodes_online.py]: Maybe I'm missing something, but where is this cache used?

File: mdp/nodes/stats_nodes_online.py [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+
+    def _check_params(self, x):
+        if self.avg is None:
+            self.avg = mdp.numx.zeros(x.shape[1], dtype=self.dtype)
+
+    def _get_supported_training_types(self):
```
Comment (nwilbert) [inline mdp/nodes/stats_nodes_online.py]: Is this method actually used? I can't remember it.

File: mdp/signal_node_online.py [hunk 16]
```diff
@@ -0,0 +1,376 @@
...
+        If numx_rng is unspecified, it will be set to a random number generator
+        with a random seed.
+        """
+        super(OnlineNode, self).__init__(input_dim, output_dim, dtype)
+        # this var stores the index of the current training iteration
+        self._train_iteration = 0
+        # this cache var dict stores an interemediate result or paramenter values
+        # at the end of each train iteration. OnlineNode subclasses should
+        # initialize the required keys
+        self._cache = dict()
```
Comment (nwilbert) [inline mdp/signal_node_online.py]: What is the advantage of having the cache dict, instead of dedicated variables in subclasses?
Maybe I missed it, but I don't see any general functionality that is enabled by this?

File: mdp/linear_flows_online.py [hunk 17]
```diff
@@ -0,0 +1,691 @@
...
+       number of nodes or a single numpy array. OnlineFlow requires only one
+       input dataiterable as each node is trained simultaneously.
+
+    c) Additional train args (supervised labels etc) are passed to each node through the
+       node specific dataiterable. OnlineFlow requires the dataiterable to return a list
+       that contains tuples of args for each node: [x, (node0 args), (node1 args), ...]. See
+       train docstring.
+
+    d) OnlineFlow also has the cache attribute that retrieves any temporary caches variables
+       stored in the OnlineNodes.
```
Comment (nwilbert) [inline mdp/linear_flows_online.py]: As you can see from the comments I created an then deleted I'm a bit confused about the benefit of the cache.
Can you describe what it is supposed to do?
Is it deeply connected to the concept of online nodes/flows? Otherwise it might be better to separate it out (possibly as an extension or by some other means).

Comment (varunrajk) [inline mdp/linear_flows_online.py]: Hi @nwilbert. When I started coding the OnlineNode, I decided to keep a dict to store a few  variables that could be useful for debugging/plotting. However as the rest of the code matured around the OnlineNode, I rarely use(d) it in the way that I had initially intended. And now after giving a second thorough look at it, I seem to agree with you. It is not actually necessary and perhaps adds redundancy to the code. In any case, if in the future if I find it really necessary, then yes, an extension could be an option. Thanks for pointing out this "vestigial" feature :). I will get rid of it and push the changes.
